### ISSUE TITLE
support for single-node multi-GPU runs using pytorch-lightning

### ISSUE THREAD

**Issue by ironv** [optuna/optuna#2333]
I am using the following [example](https://github.com/optuna/optuna/issues/1365#issuecomment-645161471)  to create code which runs on a single VM with 4 GPUs and 24 CPUS (1 thread/core).  A feed-forward network model written in pytorch-lightning is being tuned, and unlike the [example](https://github.com/optuna/optuna/blob/master/examples/pytorch_lightning_simple.py) the ranges for `nlayers`, `dropout`, etc. are set in the objective function/class. 

This code is working fine when I use only one gpu (not using the `queue` or `parallel_backend`).  When running as shown below it is running very slow and even a couple of gpus are always starved.  I tried changing `pin_memory` to `False`, `num_workers` to 2 and even 0, none of that seems to help.   When I change the backend to "multiprocessing" (like in the example), I have to set the `DataLoader` num_workers to 0, but that does not help. Any ideas?

    class FFNNPL(pl.LightningModule):
        
        ........................
        
        def train_dataloader(self):
            return utils.data.DataLoader(SparseData(smat_train, y_train)
                                        , batch_size=128, num_workers=6, shuffle=True , pin_memory=True)
        
        def val_dataloader(self):
            return utils.data.DataLoader(SparseData(smat_valid, y_valid)
                                        , batch_size=256, num_workers=6, shuffle=False, pin_memory=True)
        
        ........................
        
        
    class Objective:
        
        def __init__(self, gpu_queue):
            self.best_model_path = None
            self.trial_best_model_path = None
            self.dirpath = "checkpoints-sv3"
            
            self.gpu_queue = gpu_queue
        
        def __call__(self, trial):
            nlayers = trial.suggest_int("nlayers", 2, 5)
            dims = list(); prev_dim = 256
            for n in range(1,nlayers+1):
                prev_dim = trial.suggest_int(f"nunits_l{n}", 16, prev_dim, log=True)
                dims.append(prev_dim)
            dropout = trial.suggest_float("dropout", .4, .7, step=.02)
            lr = trial.suggest_loguniform("lr", 1E-6, 1E-3)
            params = {"model":{"input_dim":smat_train.shape[1], "hidden_dim":tuple(dims), "dropout_prob":tuple([dropout]*nlayers)}, "lr":lr}
            
            checkpoint_CB = plc.ModelCheckpoint(monitor="val_ap", save_top_k=1, mode="max", dirpath=self.dirpath
                                                        , filename=f"trial{trial.number:03}-"+"{epoch:03d}")
            earlystopping_CB = plc.early_stopping.EarlyStopping(monitor="val_ap", patience=5, mode="max")
            pruning_CB = PyTorchLightningPruningCallback(trial, monitor="val_ap")
            
            gpu_id = self.gpu_queue.get()
            trainer = pl.Trainer(weights_summary=None, gpus=[gpu_id], auto_scale_batch_size="power", deterministic=True, max_epochs=200, 
                        logger=False, progress_bar_refresh_rate=0, callbacks=[checkpoint_CB,earlystopping_CB,pruning_CB])
            trainer.fit(FFNNPL(params))
            self.trial_best_model_path = trainer.checkpoint_callback.best_model_path
            self.gpu_queue.put(gpu_id)
            
            return np.round(trainer.callback_metrics["val_ap"].item(),6)
            
            
    if __name__ == '__main__':
        ......................................................
        
        print("\nHyperparameter search", flush=True)
        pl.seed_everything(4321)
        sampler = optuna.samplers.TPESampler(seed=12345)
        pruner = optuna.pruners.MedianPruner(n_warmup_steps=10, interval_steps=2)
        study = optuna.create_study(sampler=sampler, pruner=pruner, direction="maximize")
        
        ngpu = 4
        with Manager() as manager:
            gpu_queue = manager.Queue()
            for i in range(ngpu):
                gpu_queue.put(i)
           
            with parallel_backend("threading", n_jobs=ngpu):
                obj = Objective(gpu_queue)
                study.optimize(obj, n_trials=500, n_jobs=ngpu, callbacks=[obj.callback])

**Comment by ironv** [comment]
All four GPUs are used initially. Within about a minute, one of them goes to 0 and then in about 3-4mins only one of them is being used and the rest are at 0% .  I did try with `storage=...` i.e using a db, that does not make a difference.

**Comment by anons9** [comment]
I have the same issue on my setup (4 GPUs, 128 cores CPU), where all GPUs are used and then usually get stuck or redirected to only one. Did you manage to fix it? Any hints or clues?
